### PR TITLE
[Thread]: Mark attributes which always get intercepted as External

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
What is being fixed?  Examples:
* For attributes that we always intercept, we can mark them external to save RAM, since the source of truth of those attributes are on platform, fall back to read and write from/to the attribute store does not make sense in reality. This PR only convert Thread Network Diagnostic cluster

#### Change overview
Mark attributes which always get intercepted as External

#### Testing
How was this tested? (at least one bullet point required)
* No logic change, the regression is covered by CI

### CODE REVIEW

File: src/app/zap-templates/zcl/zcl.json [hunk 1]
```diff
@@ -150,6 +150,66 @@
             "CurrentHeapHighWatermark",
             "ThreadMetrics"
         ],
+        "Thread Network Diagnostics": [
+            "channel",
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/zcl.json:154]: We can't do this, because `ThreadDiagosticsAttrAccess::Read` does fallback to the attribute store if `CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE` or `CHIP_ERROR_NOT_IMPLEMENTED` is returned....  We need to change it to not do that if we're going to add things here.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/zap-templates/zcl/zcl.json (hunk 1 above)
Comment (yufengwangca) [reply src/app/zap-templates/zcl/zcl.json:176]: acked

File: .github/workflows/tests.yaml [hunk 2]
```diff
@@ -189,7 +189,7 @@ jobs:
                   ./scripts/run_in_build_env.sh \
                   "./scripts/tests/run_test_suite.py \
                      --chip-tool ./out/darwin-x64-chip-tool${CHIP_TOOL_VARIANT}-${BUILD_VARIANT}/chip-tool \
-                     --target-skip-glob '{TestGroupMessaging}' \
+                     --target-skip-glob '{TestGroupMessaging,Test_TC_DIAG_TH_NW_1_1}' \
```
Comment (bzbarsky-apple) [inline .github/workflows/tests.yaml:192]: Is there an issue tracking this?

File: src/platform/Darwin/ThreadStackManagerImpl.cpp [hunk 3]
```diff
@@ -0,0 +1,933 @@
+/*
```
Comment (bzbarsky-apple) [inline src/platform/Darwin/ThreadStackManagerImpl.cpp:1]: Why is this code here?  It's not compiled, and can't possibly work since it's tied to various glib/Linux stuff.

The right thing to do for the YAML is probably to have it conditioned on PICS and to not claim support for this at all on Darwin if it's not supported there.

File: src/platform/Darwin/ThreadStackManagerImpl.h [hunk 4]
```diff
@@ -0,0 +1,167 @@
+/*
```
Comment (bzbarsky-apple) [inline src/platform/Darwin/ThreadStackManagerImpl.h:1]: Again, this should not exist.

File: src/platform/Linux/ThreadStackManagerImpl.cpp [hunk 5]
```diff
@@ -685,15 +685,187 @@ CHIP_ERROR ThreadStackManagerImpl::_WriteThreadNetworkDiagnosticAttributeToTlv(A
...
     case ThreadNetworkDiagnostics::Attributes::SecurityPolicy::Id:
     case ThreadNetworkDiagnostics::Attributes::OperationalDatasetComponents::Id:
-    case ThreadNetworkDiagnostics::Attributes::ActiveNetworkFaultsList::Id: {
+    case ThreadNetworkDiagnostics::Attributes::ActiveNetworkFaultsList::Id:
         err = encoder.EncodeEmptyList();
         break;
-    }
-    default: {
+    case ThreadNetworkDiagnostics::Attributes::Channel::Id:
+        err = encoder.Encode(static_cast<uint16_t>(0));
```
Comment (bzbarsky-apple) [inline src/platform/Linux/ThreadStackManagerImpl.cpp:692]: Is that a valid value?  Same for the other values here....

Note also https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/5195

File: src/platform/Linux/ThreadStackManagerImpl.cpp [hunk 6]
```diff
@@ -685,15 +685,187 @@ CHIP_ERROR ThreadStackManagerImpl::_WriteThreadNetworkDiagnosticAttributeToTlv(A
...
+        err = encoder.Encode(static_cast<uint64_t>(0));
+        break;
+    case ThreadNetworkDiagnostics::Attributes::PendingTimestamp::Id:
+        err = encoder.Encode(static_cast<uint64_t>(0));
+        break;
+    case ThreadNetworkDiagnostics::Attributes::Delay::Id:
+        err = encoder.Encode(static_cast<uint32_t>(0));
+        break;
+    case ThreadNetworkDiagnostics::Attributes::ChannelMask::Id:
+        err = encoder.Encode(ByteSpan());
```
Comment (bzbarsky-apple) [inline src/platform/Linux/ThreadStackManagerImpl.cpp:863]: _Definitely_ not a valid value; it's defined to always be 4 bytes.

**Review by bzbarsky-apple — COMMENTED** [review]
Why the submodule changes (zap, efr32_sdk)?

File: .github/workflows/tests.yaml (hunk 2 above)
Comment (yufengwangca) [reply .github/workflows/tests.yaml:192]: thread is not enabled on Darwin, why we run Thread yaml tests against this platform?

File: src/platform/Darwin/ThreadStackManagerImpl.cpp (hunk 3 above)
Comment (yufengwangca) [reply src/platform/Darwin/ThreadStackManagerImpl.cpp]: this is an accident

File: .github/workflows/tests.yaml (hunk 2 above)
Comment (bzbarsky-apple) [reply .github/workflows/tests.yaml:192]: Fair point.

File: src/platform/Linux/ThreadStackManagerImpl.cpp [hunk 7]
```diff
@@ -685,14 +688,193 @@ CHIP_ERROR ThreadStackManagerImpl::_WriteThreadNetworkDiagnosticAttributeToTlv(A
...
+        err = encoder.Encode(static_cast<uint64_t>(0));
+        break;
+    case ThreadNetworkDiagnostics::Attributes::PendingTimestamp::Id:
+        err = encoder.Encode(static_cast<uint64_t>(0));
+        break;
+    case ThreadNetworkDiagnostics::Attributes::Delay::Id:
+        err = encoder.Encode(static_cast<uint32_t>(0));
+        break;
+    case ThreadNetworkDiagnostics::Attributes::ChannelMask::Id: {
+        uint8_t buffer[sizeof(uint32_t)] = { 80, 20, 00, 00 };
```
Comment (bzbarsky-apple) [inline src/platform/Linux/ThreadStackManagerImpl.cpp:871]: Why is this the right value?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
Again, why are there submodule changes in this PR?

File: .github/workflows/darwin-tests.yaml [hunk 8]
```diff
@@ -101,7 +101,7 @@ jobs:
                   ./scripts/run_in_build_env.sh \
                   "./scripts/tests/run_test_suite.py \
                      --chip-tool ./out/darwin-x64-chip-tool-darwin-${BUILD_VARIANT}/chip-tool-darwin \
-                     --target-skip-glob '{TestGroupMessaging}' \
+                     --target-skip-glob '{TestGroupMessaging,Test_TC_DIAG_TH_NW_1_1,Test_TC_DIAG_TH_NW_1_2}' \
```
Comment (bzbarsky-apple) [inline .github/workflows/darwin-tests.yaml:104]: The right place to disable these for chip-tool-darwin is probably `examples/chip-tool-darwin/templates/tests/tests.js` instead of here.

File: src/app/tests/suites/certification/Test_TC_DIAG_TH_NW_1_2.yaml [hunk 9]
```diff
@@ -109,13 +95,6 @@ tests:
               - name: "expectedValue"
                 value: "y"
 
-    - label: "Validate constraints of attribute: ExtendedPanId"
```
[the remark this answers is not in the document]
Comment (yufengwangca) [reply src/app/tests/suites/certification/Test_TC_DIAG_TH_NW_1_2.yaml]: Added back to verify against new type to align with spec

File: .github/workflows/darwin-tests.yaml (hunk 8 above)
Comment (yufengwangca) [reply .github/workflows/darwin-tests.yaml]: it looks we need to change both places to pass CI

Comment (bzbarsky-apple) [reply .github/workflows/darwin-tests.yaml]: Why?  this is going to run "all the tests chip-tool-darwin knows about", and that JS file can be used to change the set of tests it knows about....

File: .github/workflows/darwin-tests.yaml [hunk 10]
```diff
@@ -101,7 +101,7 @@ jobs:
                   ./scripts/run_in_build_env.sh \
                   "./scripts/tests/run_test_suite.py \
                      --chip-tool ./out/darwin-x64-chip-tool-darwin-${BUILD_VARIANT}/chip-tool-darwin \
-                     --target-skip-glob '{TestGroupMessaging}' \
+                     --target-skip-glob '{TestGroupMessaging,Test_TC_DIAG_TH_NW_1_1,Test_TC_DIAG_TH_NW_1_2,Test_TC_DIAG_TH_NW_2_2,Test_TC_DIAG_TH_NW_2_3,Test_TC_DIAG_TH_NW_2_6,Test_TC_DIAG_TH_NW_2_7,Test_TC_DIAG_TH_NW_2_8,Test_TC_DIAG_TH_NW_2_9}' \
```
Comment (bzbarsky-apple) [inline .github/workflows/darwin-tests.yaml:104]: Again, this is not the right change.  These should be disabled, with a comment explaining why, in examples/chip-tool-darwin/templates/tests/tests.js.

File: src/app/tests/suites/certification/Test_TC_DIAG_TH_NW_1_2.yaml [hunk 11]
```diff
@@ -28,36 +28,30 @@ tests:
...
-          value: 0
+          value: null
 
     - label: "Validate constraints of attribute: Channel"
       command: "readAttribute"
       attribute: "channel"
       response:
           constraints:
-              type: uint16
+              type: null
```
Comment (bzbarsky-apple) [inline src/app/tests/suites/certification/Test_TC_DIAG_TH_NW_1_2.yaml:42]: This does not make sense.  The type is uint16 (but nullable).  `null` is not a type.  Please undo this change.

File: src/app/tests/suites/certification/Test_TC_DIAG_TH_NW_1_2.yaml [hunk 12]
```diff
@@ -114,7 +107,7 @@ tests:
       attribute: "ExtendedPanId"
       response:
           constraints:
-              type: uint64
+              type: null
```
Comment (bzbarsky-apple) [inline src/app/tests/suites/certification/Test_TC_DIAG_TH_NW_1_2.yaml:110]: Again, does not make sense.

File: src/app/tests/suites/certification/Test_TC_DIAG_TH_NW_1_2.yaml [hunk 13]
```diff
@@ -248,35 +241,35 @@ tests:
       attribute: "PartitionId"
       response:
           constraints:
-              type: uint32
+              type: null
```
Comment (bzbarsky-apple) [inline src/app/tests/suites/certification/Test_TC_DIAG_TH_NW_1_2.yaml:244]: Does not make sense.  I'm going to stop commenting on these, but there are a lot....  Please undo all these contraint type changes.

File: src/platform/Linux/ThreadStackManagerImpl.cpp [hunk 14]
```diff
@@ -685,15 +685,169 @@ CHIP_ERROR ThreadStackManagerImpl::_WriteThreadNetworkDiagnosticAttributeToTlv(A
...
+        err = encoder.Encode(static_cast<uint32_t>(0));
+        break;
+    case ThreadNetworkDiagnostics::Attributes::RxErrFcsCount::Id:
+        err = encoder.Encode(static_cast<uint32_t>(0));
+        break;
+    case ThreadNetworkDiagnostics::Attributes::RxErrOtherCount::Id:
+        err = encoder.Encode(static_cast<uint32_t>(0));
+        break;
+    case ThreadNetworkDiagnostics::Attributes::ActiveTimestamp::Id:
+        err = encoder.Encode(static_cast<uint64_t>(0));
```
Comment (bzbarsky-apple) [inline src/platform/Linux/ThreadStackManagerImpl.cpp:834]: Why is this not null?  Seriously, please check this against the spec.  There are more things here that don't match....

File: src/platform/Linux/ThreadStackManagerImpl.cpp [hunk 15]
```diff
@@ -685,15 +685,169 @@ CHIP_ERROR ThreadStackManagerImpl::_WriteThreadNetworkDiagnosticAttributeToTlv(A
     case ThreadNetworkDiagnostics::Attributes::RouteTableList::Id:
     case ThreadNetworkDiagnostics::Attributes::SecurityPolicy::Id:
     case ThreadNetworkDiagnostics::Attributes::OperationalDatasetComponents::Id:
```
Comment (bzbarsky-apple) [inline src/platform/Linux/ThreadStackManagerImpl.cpp:687]: Why is OperationalDatasetComponents being encoded as a list?  Pre-existing, so separate PR is probably OK.

Comment (yufengwangca) [reply src/platform/Linux/ThreadStackManagerImpl.cpp]: https://github.com/project-chip/connectedhomeip/issues/18906

File: .github/workflows/darwin-tests.yaml (hunk 8 above)
Comment (yufengwangca) [reply .github/workflows/darwin-tests.yaml]: Because I hit a couple of issues doing this.

After I disable Test_TC_DIAG_TH_NW_1_1 in examples/chip-tool-darwin/templates/tests/tests.js, still failed in Test Suites - Darwin (See the CI failure) 

If I specify Test_TC_DIAG_TH_NW_2_6 ~ Test_TC_DIAG_TH_NW_2_9 in examples/chip-tool-darwin/templates/tests/tests.js, I won't event pass the codegen.

```
🤖 Generating files:
    ✍  /__w/connectedhomeip/connectedhomeip/zzz_generated/chip-tool-darwin/zap-generated/cluster/CHIPTestClustersObjc.h
    ✍  /__w/connectedhomeip/connectedhomeip/zzz_generated/chip-tool-darwin/zap-generated/cluster/CHIPTestClustersObjc.mm
⚠️  Errors:
    👎  /__w/connectedhomeip/connectedhomeip/zzz_generated/chip-tool-darwin/zap-generated/test/Commands.h: ⛔ Error: Test Test_TC_DIAG_TH_NW_2_6  does not exists. [line: 19, column: 22, file: /__w/connectedhomeip/connectedhomeip/examples/chip-tool-darwin/templates/tests/commands.zapt ]
Stack trace:
```

**Review by bzbarsky-apple — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]
